service/gcs/core/gcs/gcs.go

```diff
@@ -468,35 +471,29 @@ func (c *gcsCore) RunExternalProcess(params prot.ProcessParameters, stdioSet *co
 		}
 		utils.LogMsgf("external process %d exited with exit status %d", cmd.Process().Pid(), cmd.ExitState().ExitCode())
 
-		// Close stdin so that the copying goroutine is safely unblocked; this is necessary
-		// because the host expects stdin to be closed before it will report process
-		// exit back to the client, and the client expects the process notification before
-		// it will close its side of stdin (which io.Copy is waiting on in the copying goroutine).
-		if stdioSet.In != nil {
-			if err := stdioSet.In.CloseRead(); err != nil {
-				logrus.Errorf("failed call to CloseRead for external process stdin: %v: %s", ociProcess.Args, err)
+		if master != nil {
```

Comment:
With the new code, we close stdin after we launch the process (the defers handle this), so there are no copy operations anymore that could block. So when the child process closes stdin, the host gets the notification directly.